src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,549 @@
+        /// is global namespace, i.e. default namespace is "" and the file is located at project 
+        /// root directory, and no namespace declaration in the document, respectively.
+        /// </summary>
+        internal sealed class ChangeNamespaceCodeAction : CodeAction
+        {
+            private readonly State _state;
+            private readonly AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax, TMemberDeclarationSyntax> _service;
+            private readonly ImmutableArray<string> _oldNamespaceParts;
+            private readonly ImmutableArray<string> _newNamespaceParts;
+
+            public override string Title => _state.TargetNamespace.Length == 0 
+                ? FeaturesResources.Change_to_global_namespace
+                : string.Format(FeaturesResources.Change_namespace_to_0, _state.TargetNamespace);
+
+            public ChangeNamespaceCodeAction(
+                AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax, TMemberDeclarationSyntax> service,
+                State state)
+            {
+                Debug.Assert(state.TargetNamespace != null);
+
+                _service = service;
+                _state = state;
+
+                var dotSeparator = new[] { '.' };
+                _oldNamespaceParts = _state.DeclaredNamespace.Split(dotSeparator).ToImmutableArray();
+                _newNamespaceParts = _state.TargetNamespace.Split(dotSeparator).ToImmutableArray();
+            }
+
+            private ImmutableArray<ISymbol> GetDeclaredSymbolsInContainer(
+                SemanticModel semanticModel, 
+                SyntaxNode node, 
+                CancellationToken cancellationToken)
+            {
+                var declarations = _service.GetMemberDeclarationsInContainer(node);
+                var builder = ArrayBuilder<ISymbol>.GetInstance();
+                foreach (var declaration in declarations)
+                {
+                    var symbol = semanticModel.GetDeclaredSymbol(declaration, cancellationToken);
+                    builder.AddIfNotNull(symbol);
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            protected override async Task<Solution> GetChangedSolutionAsync(CancellationToken cancellationToken)
+            {
+                // Here's the entire process for changing namespace:
+                // 1. Change the namespace declaration, fix references and add imports that might be necessary.
+                // 2. Explicitly merge the diff to get a new solution.
+                // 3. Remove added imports that are unnecessary.
+                // 4. Do another explicit diff merge based on last merged solution.
+                //
+                // The reason for doing explicit diff merge twice is so merging after remove unnecessaty imports can be correctly handled.
+
+                var solutionAfterNamespaceChange = _state.Solution;
```

Comment:
I see , so fix is not lazy? but everything is done up front and just to applied to solution?

Reply:
Not sure I understand what do you mean by "lazy", I believe this method `GetChangedSolutionAsync` won't be called until user select the refactoring and we have to show the preview